### PR TITLE
TP-563  Improve tool performance

### PR DESCRIPTION
**Pull request by MohamedFarouk-HMCTS**

### CODE REVIEW

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java [hunk 1]
```diff
@@ -1,78 +1,104 @@
...
+
+    private void generateDataIfRequired() throws InvalidPositionException, SQLException, LiquibaseException {
+        final boolean generateTestData = valueOf(getProperty("generateTestData", GENERATE_TEST_DATA));
+        final long eventsStreamCount = Long.valueOf(getProperty("eventsStreamCount"));
+        final long eventsPerStreamCount = Integer.valueOf(getProperty("eventsPerStreamCount"));
+        final long testDataCountStepInfo = Integer.valueOf(getProperty("testDataCountStepInfo"));
+
+        if (generateTestData) {
+            if (eventsPerStreamCount == 0l || eventsStreamCount == 0l) {
+                LOGGER.info("================================================================================================\n" +
```
Comment (matt-rich) [inline stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: maybe put these types of messages in constant to make the code cleaner.

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java [hunk 2]
```diff
@@ -1,78 +1,104 @@
...
+        LOGGER.info("================================================================================================");
 
-        for (final UUID id : streamIds) {
-            for (long logId = 1; logId <= EVENTS_PER_STREAM; logId++) {
-                databaseUtils.getEventLogJdbcRepository().insert(eventLogFrom("sample.events.name", logId, id, createdAt));
+    }
+
+    private void generateDataIfRequired() throws InvalidPositionException, SQLException, LiquibaseException {
+        final boolean generateTestData = valueOf(getProperty("generateTestData", GENERATE_TEST_DATA));
+        final long eventsStreamCount = Long.valueOf(getProperty("eventsStreamCount"));
```
Comment (matt-rich) [inline stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: can't we only extract these additional values within the IF condition below

File: event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java [hunk 3]
```diff
@@ -25,14 +25,20 @@
...
+import org.wildfly.swarm.spi.runtime.annotations.ConfigurationValue;
 
 @Singleton
 @Startup
 public class StartTransformation implements ManagedTaskListener {
 
     private static final String NO_PROCESS_FILE_WARNING = "!!!!! No Swarm Process File specific, application will not auto-shutdown on completion. Please use option '-Dorg.wildfly.swarm.mainProcessFile=/pathTo/aFile' to specify location of process file with read/write permissions !!!!!";
 
+    @Inject
+    @ConfigurationValue("streams-processed-count-step-info")
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java]: yes it is, seems like a better name so changed to what you have mentioned

File: event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java [hunk 4]
```diff
@@ -52,10 +58,16 @@
 
     boolean allTasksCreated = false;
 
+    private StopWatch stopWatch = new StopWatch();
+
+    private AtomicInteger streamsDoneCount = new AtomicInteger(0);
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java]: processedStreamsCount

File: event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java [hunk 5]
```diff
@@ -65,12 +77,22 @@ void go() {
 
     private void createTransformationTasks(final int pass) {
 
+        if (streamsProcessedCountStepInfo == 0) {
+            throw new IllegalArgumentException("Invalid streamsProcessedCountStepInfo argument");
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java]: Wiki explains this and they would get error if they dont set a config value

File: event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java [hunk 6]
```diff
@@ -65,12 +77,22 @@ void go() {
...
+        }
+
         final Stream<UUID> activeStreams = eventRepository.getAllActiveStreamIds();
 
         activeStreams
                 .forEach(streamId -> {
                     final StreamTransformationTask transformationTask = new StreamTransformationTask(streamId, eventStreamTransformationService, this, pass);
                     outstandingTasks.add(executorService.submit(transformationTask));
+                    streamsDoneCount.getAndIncrement();
+                    if ((streamsDoneCount.get() != 0 && streamsProcessedCountStepInfo != 0)
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java]: done

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/SwarmStarterUtil.java [hunk 7]
```diff
@@ -19,9 +19,12 @@
 
     private static final Logger LOGGER = getLogger(SwarmStarterUtil.class);
 
-    public void runCommand(final boolean enableRemoteDebugging, final long timeoutInSeconds) throws IOException {
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/SwarmStarterUtil.java]: This is all documented in the WIKI how to use these parameters
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-failsafe-plugin</artifactId>
                <configuration>
                    <systemPropertyVariables>
                        <generateTestData>true</generateTestData>
                        <eventsStreamCount>10</eventsStreamCount>
                        <eventsPerStreamCount>10</eventsPerStreamCount>
                        <databaseEventStreamCount>10</databaseEventStreamCount>
                        <databaseEventLogCount>100</databaseEventLogCount>
                        <runTest>true</runTest>
                        <streamCountReportingInterval>2</streamCountReportingInterval>
                        <generatedTestDataReportingInterval>10</generatedTestDataReportingInterval>
                        <debug>false</debug>
                        <memoryOptions>2048</memoryOptions>
                    </systemPropertyVariables>
                    <argLine></argLine>
                </configuration>
            </plugin>

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/GenerateTestData.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.UUID;
+
+import liquibase.exception.LiquibaseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GenerateTestData {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GenerateTestData.class);
+
+    private static final String STREAMS_COUNT = "100";
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/GenerateTestData.java]: added

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java (hunk 2 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: fixed

File: stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationIT.java [hunk 9]
```diff
@@ -47,7 +50,7 @@ public void shouldTransformEventInEventStore() throws Exception {
         databaseUtils.insertEventLogData("sample.events.name", STREAM_ID, 1L, createdAt);
         databaseUtils.insertEventLogData("sample.v2.events.name", STREAM_ID, 2L, createdAt);
 
-        swarmStarterUtil.runCommand(ENABLE_REMOTE_DEBUGGING_FOR_WILDFLY, WILDFLY_TIMEOUT_IN_SECONDS);
+        swarmStarterUtil.runCommand(STREAMS_PROCESSED_COUNT_STEP_INFO, ENABLE_REMOTE_DEBUGGING_FOR_WILDFLY, WILDFLY_TIMEOUT_IN_SECONDS, MEMORY_OPTIONS);
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationIT.java]: Done the refactoring

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java (hunk 1 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: DOne

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/GenerateTestData.java [hunk 10]
```diff
@@ -18,17 +18,25 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+/**
+ * This class by default creates 1000 events with 100 streams with 10 events per stream These can be
```
Comment (matt-rich) [inline stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/GenerateTestData.java]: you could potentially use the @link #variableName to point to the 100, 10 events so it's clear where to adjust.

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java [hunk 11]
```diff
@@ -19,6 +19,9 @@
 
 public class StreamTransformationPerformanceIT {
     private static final Logger LOGGER = LoggerFactory.getLogger(StreamTransformationPerformanceIT.class);
+    final String LINE_BREAK = "================================================================================================\n";
```
Comment (matt-rich) [inline stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: private static ?

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java [hunk 12]
```diff
@@ -36,14 +39,16 @@ public void setupDatabase() throws Exception {
 
     }
 
-    @Ignore("Disabled by default, remove Ignore to run performance test")
+    //@Ignore("Disabled by default, remove Ignore to run performance test")
```
Comment (matt-rich) [inline stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: do we want this ignored or commented out?

Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: ignored for IT test  removed the //

File: stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java (hunk 11 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-performance-test/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationPerformanceIT.java]: fixed

**Pull request merged**